File: Dart/Longest Substring Without Repeating Characters.dart
import 'dart:math';

class Solution {
  int lengthOfLongestSubstring(String s) {
    int maxSubString = 0;
    Set<String> set = Set();
    int start = 0, end = 0;
    while (end < s.length) {
      if (!set.contains(s[end])) {
        set.add(s[end]);
        maxSubString = max(maxSubString, set.length);
        end++;
      } else {
        while (s[start] != s[end]) {
          set.remove(s[start]);
          start++;
        }
        set.remove(s[start]);
        start++;
      }
    }
    return maxSubString;
  }
}

void main() {
  Solution solution = Solution();
  String testString = "abcbb";
  int result = solution.lengthOfLongestSubstring(testString);
  print(
      "The length of the longest substring without repeating characters is: $result");
}

File: Python/longest-substring-without-repeating-characters.py
from typing import Set


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0
        start = 0
        len_s = len(s)
        s_set: Set[str] = set([s[0]])
        result = 1

        for end in range(1, len_s):
            while s[end] in s_set:
                s_set.remove(s[start])
                start += 1
            result = max(result, end - start + 1)
            s_set.add(s[end])
        return result


if __name__ == "__main__":
    s = "abcabcbb"
    sol = Solution()
    print(sol.lengthOfLongestSubstring(s))
